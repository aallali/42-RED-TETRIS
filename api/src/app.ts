
import * as express from "express";
import * as cors from "cors"
import { Socket } from "socket.io";
import Game from "./Game"

import ROOM from "./Room"
import PLAYER from "./Player"
import { ROOM_MODE } from "./types"


const log = console.log
interface IPlayer {
	id: string;
	nickname: string;
}
interface IRoom {
	id: string;
	name: string;
	players: []
	admin: IPlayer
}
class App {

	public app: express.Application;
	public http: any
	public io: Socket
	private port: number = 4242
	private gameIns: Game
	constructor() {
		this.app = express();
		this.app.use(
			cors({
				origin: "*",
				credentials: true,
			})
		);
		this.http = require("http").Server(this.app);
		this.io = require("socket.io")(this.http, {
			pingInterval: 60000,
			cors: {
				origin: "*",
			},
		});
		this.routes()
		this.initSocketEvents()
		this.gameIns = new Game(this.io)

	}
	/**
	 * 
	 */
	private routes() {
		this.app.get("/", (req: any, res: any) => {
			console.log(Array.from(this.gameIns.players))
			res.send(Array.from(this.gameIns.players))
		});
	}
	private initSocketEvents() {
	}
	public listen() {
		console.clear()
		this.http.listen(this.port, () =>
			console.log(`ðŸš€ Server is running on PORT ${this.port}`)
		);
	}
}

export default new App();
