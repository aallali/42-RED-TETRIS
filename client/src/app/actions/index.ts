import {
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER, UPDATE_SCORE,
	SET_PLAYER_ADMIN, ADD_ROW,
	PLAYER_LOST, RESET_STATES,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN
} from "../../reducers/player.reducer";
import { UPDATE_PLAYERS, CLEAR_OPPONENTS } from "../../reducers/opponent.reducer"
import { SET_ERROR, CLEAR_ERROR } from "../../reducers/error.reducer"

import {
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
} from "../../reducers/game.reducer"

export {
	// Players actions
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER,
	UPDATE_SCORE,
	SET_PLAYER_ADMIN,
	ADD_ROW, PLAYER_LOST,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN,
	RESET_STATES,

	// Opponents actions
	UPDATE_PLAYERS,
	CLEAR_OPPONENTS,
	// Error actions
	SET_ERROR, CLEAR_ERROR,

	// game actions
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
}
