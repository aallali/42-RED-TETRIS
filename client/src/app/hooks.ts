import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { useGameStatus } from "./TetrisHooks/useGameStatus"
import { useInterval } from "./TetrisHooks/useInterval"
import { usePlayer } from "./TetrisHooks/usePlayer"
import { useStage } from "./TetrisHooks/useStage"


import io from "socket.io-client";

export const socket = io("http://localhost:4242");
export { useGameStatus, usePlayer, useInterval, useStage }
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
