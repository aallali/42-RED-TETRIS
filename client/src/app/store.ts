import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import playerReducer from "../reducers/player.reducer";
import stageReducer from "../reducers/stage.reducer"
export const store = configureStore({
  reducer: {
    player: playerReducer,
    stage: stageReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
