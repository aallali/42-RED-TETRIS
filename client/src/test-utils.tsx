// test-utils.jsx
import React, { ReactChildren } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import playerReducer from "./reducers/player.reducer";
import opponentReducer from "./reducers/opponent.reducer"
import errorReducer from "./reducers/error.reducer"
import gameReducer from "./reducers/game.reducer"
import chatReducer from './reducers/chat.reducer';

function aallaliStore(preloadedState: any) {
	return configureStore({
		reducer: {
			player: playerReducer,
			opponents: opponentReducer,
			error: errorReducer,
			game: gameReducer,
			chat: chatReducer
		},
		preloadedState
	})
}
function render(
	ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
	{
		preloadedState,
		store = configureStore({
			reducer: {
				player: playerReducer,
				opponents: opponentReducer,
				error: errorReducer,
				game: gameReducer,
				chat: chatReducer
			},
			preloadedState
		}),
		...renderOptions
	}: any = {}
) {
	function Wrapper({ children }: { children: ReactChildren }) {
		return <Provider store={store}>{children}</Provider>

	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}


// re-export everything
export * from '@testing-library/react'
// override render method
export { render, aallaliStore }