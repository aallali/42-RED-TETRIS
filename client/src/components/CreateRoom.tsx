/* eslint-disable react-hooks/exhaustive-deps */

import { ChangeEvent, useState } from "react";
import { UPDATE_GAME_MODE } from "../app/actions";
import { useAppDispatch, useAppSelector } from "../app/hooks";
import { getGameMode } from "../reducers/game.reducer";
import { getPlayerNickname } from "../reducers/player.reducer";

export default function CreateRoomInput() {
	const game_mode = useAppSelector(getGameMode)
	const playerName = useAppSelector(getPlayerNickname)
	const [roomName, setRoomName] = useState<string>("")
	const [error, setError] = useState<string>("")

	const dispatch = useAppDispatch()
	function handleGameModeChange(event: ChangeEvent<HTMLInputElement>) {
		if (event.target.checked)
			dispatch(UPDATE_GAME_MODE("multiplayer"))
		else
			dispatch(UPDATE_GAME_MODE("solo"))
	}

	function handleCreateRoom(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
		e.preventDefault();
		if (roomName && /^[a-zA-Z0-9]{1,10}$/.test(roomName))
			window.location.hash = `#${roomName}[${playerName}]`
		else
			setError("alphabetic/numbers only (length between 1<x<10)")
	}
	return (
		<div
			className="flex my-1  rounded-sm pt-3 pb-3 pr-4 pl-4  bg-blue-50 justify-between items-center">
			<div className="flex flex-justify-start gap-1 items-center">
				<div
					className="relative inline-block w-10 mr-1 align-middle select-none transition duration-200 ease-in">
					<input type="checkbox" name="toggle" id="toggle" value={game_mode}
					data-testid="toggle"
						className="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
						onChange={handleGameModeChange} defaultChecked={game_mode === "multiplayer" ? true : false}
					/>
					<label htmlFor="toggle"
						className="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"></label>
				</div>

				<span className="uppercase font-bold text-gray-700">
					<label htmlFor="toggle" className="">{game_mode}</label>
				</span>
			</div>
			<div>
				<form className="m-0 flex">
				<input
					className="text-xl w-full rounded-l-md p-2 border-t mr-0 border-b border-l text-gray-800 border-gray-200 bg-white"
					placeholder="room name..."
					value={roomName}
					id="roomNameInput"
					onChange={(e) => { setRoomName(e.target.value); }} />

				<button
					className="text-md w-1/2 font-bold px-1 rounded-r-md bg-blue-400  text-gray-800 font-bold p-1 uppercase border-none" onClick={handleCreateRoom}>Create
				</button>
			</form>
				<p className="text-xs text-red-600">{error}</p>
			</div>

		</div>

	);
}
