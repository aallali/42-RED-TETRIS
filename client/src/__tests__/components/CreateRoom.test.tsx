import renderer from "react-test-renderer";
import CreateRoom from "../../components/CreateRoom"
import { Provider } from 'react-redux';
import * as reactRedux from "react-redux";
import { render, fireEvent, screen } from '../../test-utils'
import { store } from "../../app/store"
// so we can import fireEvent and screen here as well

jest.setTimeout(10 * 1000)

describe("_ ChatBox Component test", () => {
	const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
	beforeEach(() => {
		useDispatchMock.mockClear();
	});

	test("test chatbox rendering ", () => {
		const tree = renderer.create(<Provider store={store}>
			<CreateRoom />
		</Provider>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	test('test send message chatbox', async () => {
		useDispatchMock.mockReturnValue(jest.fn());

		render(
			<CreateRoom />,
			{ game: { mode: "solo" } }
		)

		const checkbox = screen.getByTestId("toggle")
		const createRoomInput = screen.getByPlaceholderText("room name...")

		fireEvent.change(createRoomInput, { target: { value: "" } });
		fireEvent.click(screen.getByText(/CREATE/i));
		expect(screen.getByText(/alphabetic\/numbers only/)).toBeInTheDocument()

		fireEvent.change(createRoomInput, { target: { value: "arena" } });
		fireEvent.click(screen.getByText(/CREATE/i));
		expect(useDispatchMock).toHaveBeenCalled();

		// fireEvent.change(checkbox, { target: { checked: true, value: "multiplayer" } });
		fireEvent.click(checkbox);
		expect(screen.getByText(/multiplayer/i)).toBeInTheDocument()

		render(
			<CreateRoom />,
			{
				preloadedState:
				{
					game: { mode: "solo" }
				}
			}
		)

		// fireEvent.change(checkbox, { target: { checked: false, value: "solo" } });
		fireEvent.click(checkbox);
		expect(screen.getByText(/solo/i)).toBeInTheDocument()
		// fireEvent.change(checkbox, { target: { checked: true } });

	})
})