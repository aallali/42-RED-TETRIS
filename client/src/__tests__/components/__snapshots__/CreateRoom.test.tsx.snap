// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_ ChatBox Component test test chatbox rendering  1`] = `
<div
  className="flex my-1  rounded-sm pt-3 pb-3 pr-4 pl-4  bg-blue-50 justify-between items-center"
>
  <div
    className="flex flex-justify-start gap-1 items-center"
  >
    <div
      className="relative inline-block w-10 mr-1 align-middle select-none transition duration-200 ease-in"
    >
      <input
        className="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
        data-testid="toggle"
        defaultChecked={true}
        id="toggle"
        name="toggle"
        onChange={[Function]}
        type="checkbox"
        value="multiplayer"
      />
      <label
        className="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"
        htmlFor="toggle"
      />
    </div>
    <span
      className="uppercase font-bold text-gray-700"
    >
      <label
        className=""
        htmlFor="toggle"
      >
        multiplayer
      </label>
    </span>
  </div>
  <div>
    <form
      className="m-0 flex"
    >
      <input
        className="text-xl w-full rounded-l-md p-2 border-t mr-0 border-b border-l text-gray-800 border-gray-200 bg-white"
        id="roomNameInput"
        onChange={[Function]}
        placeholder="room name..."
        value=""
      />
      <button
        className="text-md w-1/2 font-bold px-1 rounded-r-md bg-blue-400  text-gray-800 font-bold p-1 uppercase border-none"
        onClick={[Function]}
      >
        Create
      </button>
    </form>
    <p
      className="text-xs text-red-600"
    >
      
    </p>
  </div>
</div>
`;
